js_modules/dagit/packages/core/src/workspace/WorkspaceJobsRoot.tsx

```diff
@@ -0,0 +1,369 @@
+      if (entry.locationOrLoadError?.__typename !== 'RepositoryLocation') {
+        return;
+      }
+
+      const location = entry.locationOrLoadError;
+      entry.locationOrLoadError.repositories.forEach((repo) => {
+        const repoAddress = buildRepoAddress(repo.name, location.name);
+        repo.schedules.forEach((schedule) => {
+          const jobKey = makeJobKey(repoAddress, schedule.pipelineName);
+          const dataForJob = {...(byJobKey[jobKey] || {schedules: [], sensors: []})};
+          dataForJob.schedules.push(schedule);
+          byJobKey[jobKey] = dataForJob;
+        });
+        repo.sensors.forEach((sensor) => {
+          (sensor?.targets || []).forEach((target) => {
+            const jobKey = makeJobKey(repoAddress, target.pipelineName);
+            const dataForJob = {...(byJobKey[jobKey] || {schedules: [], sensors: []})};
+            dataForJob.sensors.push(sensor);
+            byJobKey[jobKey] = dataForJob;
+          });
+        });
+      });
+    });
+
+    return byJobKey;
+  }, [data]);
+
+  const runsByJob: {[key: string]: RunTimeFragment[]} = React.useMemo(() => {
+    if (!recentRunsData || recentRunsData?.workspaceOrError.__typename !== 'Workspace') {
+      return {};
+    }
+
+    const byJobKey = {};
+    recentRunsData.workspaceOrError.locationEntries.forEach((entry) => {
+      if (entry.locationOrLoadError?.__typename !== 'RepositoryLocation') {
+        return;
+      }
+      const location = entry.locationOrLoadError;
+      entry.locationOrLoadError.repositories.forEach((repo) => {
+        const repoAddress = buildRepoAddress(repo.name, location.name);
+        repo.pipelines.forEach((job) => {
+          const jobKey = makeJobKey(repoAddress, job.name);
+          byJobKey[jobKey] = job.runs;
+        });
+      });
+    });
+
+    return byJobKey;
+  }, [recentRunsData]);
+
+  const repoBuckets = React.useMemo(() => {
+    return [...allRepos]
+      .sort((a, b) =>
+        a.repository.name.toLocaleLowerCase().localeCompare(b.repository.name.toLocaleLowerCase()),
+      )
+      .map((repo) => {
+        const {name, pipelines} = repo.repository;
+        const repoAddress = buildRepoAddress(name, repo.repositoryLocation.name);
+        return {
+          repoAddress,
+          jobs: pipelines
+            .filter(({name}) => !isHiddenAssetGroupJob(name))
+            .map((pipeline) => {
+              const jobKey = makeJobKey(repoAddress, pipeline.name);
+              const dataForJob = schedulesAndSensorsByJob[jobKey];
+
+              return {
+                isJob: pipeline.isJob,
+                name: pipeline.name,
+                repoAddress,
+                schedules: dataForJob?.schedules || [],
+                sensors: dataForJob?.sensors || [],
+                runs: runsByJob[jobKey] || [],
+              };
+            }),
+        };
+      })
+      .filter((repo) => repo.jobs.length > 0);
+  }, [allRepos, runsByJob, schedulesAndSensorsByJob]);
```

Comment:
Hmm, it's a bit hard to get the "gist" of what's going on in this component because of the amount of data transformation that happens here. I really like that you've broken out things like filtering by search string into separate steps below because it's easy to spot those.

I wonder if it'd make sense to move some of the data fetching or these useMemo's out into named hooks. They wouldn't be reused anywhere, but it might be nice to say "useSchedulesAndSensorsByJob" and know that nothing exciting was happening in there and you could safely skip it  😅

Reply:
I can't move the data fetching into these hooks since I think we will want to have access to them in a refresh interval in the main component. But hopefully this is a little easier to follow.